File: comb/#39_Combination_Sum.py
# Can be optimized more, currently O(2^n)
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        arr = []
        
        def helper(i, comb_sum):
            # Base cases
            if comb_sum == target:
                res.append(arr.copy())
                return
            
            if i >= len(nums) or comb_sum > target:
                return
            
            # Choice 1: Include current element (stay at same index for reuse)
            arr.append(nums[i])
            helper(i, comb_sum + nums[i])  # Stay at i to allow reuse
            arr.pop()  # Backtrack
            
            # Choice 2: Skip current element (move to next index)
            helper(i + 1, comb_sum)
        
        helper(0, 0)
        return res

File: Leetcode/Cpp/#39_Combination_Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> curr;
        sort(candidates.begin(), candidates.end());
        helper(candidates, target, curr, res, 0);
        return res;
    }
    void helper(vector<int>& candidates, int target, vector<int>& curr, vector<vector<int>>& res, int i) {
        if(target < 0) return;
        if(target == 0) {
            res.push_back(curr);
            return;
        }
        for(int j = i; j < candidates.size(); j++) {
            curr.push_back(candidates[j]);
            helper(candidates, target - candidates[j], curr, res, j);
            curr.pop_back();
        }

    }
};
